### ISSUE TITLE
Running Oqtane on a load balancer

### ISSUE THREAD

**Issue by mixlplix** [oqtane/oqtane.framework#973]
I am trying to get Oqtane running in a production environment.  This is on AWS with a load balancer that is terminating HTTPS.  I have been troubleshooting this for several days now.  

The public url is: https://oqtane.cmrus.com/

I have tried various things like disabling https redirection, adding header forwarding and none of them seem to work.  I created a standard Blazor Server app (File...New Project), and published the app to the same server location and the Blazor server app works just fine. When I hit the URL internally from the machine, or over the VPN, the site also loads correctly.  I am using a vanilla install of Oqtane.

Any suggestions?

**Comment by sbwalker** [comment]
When you installed Oqtane did you use the Install Wizard to provide the database connection details, or did you not even get that far?

**Comment by mixlplix** [comment]
I published Oqtane from Visual Studio using a network share over the VPN.

**Comment by sbwalker** [comment]
In the web console I can see that the client is trying to call the server:

at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
at Oqtane.Services.ServiceBase.GetJsonAsync[T](String uri) in C:\code\blazor\oqtane.framework\Oqtane.Client\Services\ServiceBase.cs:line 57
at Oqtane.Services.InstallationService.IsInstalled() in C:\code\blazor\oqtane.framework\Oqtane.Client\Services\InstallationService.cs:line 16
at Oqtane.App.OnParametersSetAsync() in C:\code\blazor\oqtane.framework\Oqtane.Client\App.razor:line 27

It would be trying to call the Url: https://oqtane.cmrus.com/~/api/Installation/installed but it is throwing a 503 Service Unavailable

I tried to hit the Swagger UI endpoint as well at https://oqtane.cmrus.com/swagger/index.html and it is throwing a 503 Service Unavailable

My assumption is that the Visual Studio publish did not work properly. We have instructions on how to install Oqtane on IIS here:  https://www.oqtane.org/Resources/Blog/PostId/542/installing-oqtane-on-iis

**Comment by mixlplix** [comment]
True, it looks like I did not install Oqtane, When I go to the root address on the local machine I see the installation screen, and I clicked “Install Now”, but am getting an error about local runtime db, looks like I need to enable something on that machine.  The swagger URL works now, though.

 

https://oqtane.cmrus.com/swagger/index.html

 

So https traffic is being forwarded correctly, so I would think I should be able to get to the install screen by going here, though.

 

https://oqtane.cmrus.com/

 

Because I can locally.  We are looking into enabling wss:// at the load balancer.  If you look in the developer console you will see it trying to connect to 

 

wss://oqtane.cmrus.com/_blazor?id=bIsxxa2UxFm75uTJDXmOjA 

 

for the web socket, and we are thinking the reply is getting lost….?

 

 

</doug>

**Comment by sbwalker** [comment]
If you are running on Blazor Server then you are using SignalR under the covers. So you would need to make sure that web sockets are configured to work with your web farm configuration. Another thing to keep in mind is when you deployed the default Blazor Server template app, it is a server app and is not architected in a true client/server manner. It does not have any data access so it’s not really a real world Blazor app. Oqtane uses a client/server architecture - so regardless of whether you are running on Blazor Server or Blazor WebAssembly, the client application is making HTTP calls to the Web API in the server app. It appears that this is where your web farm setup is having difficulty - the calls from the client are not reaching the server because the server APIs are not accessible. Is there a reason why you chose to use Blazor Server?

**Comment by markdav-is** [comment]
Hi, I'm on @mixlplix 's team.  We are using Blazor Server version of Oqtane because we are trying to use it for a high-volume ticketing system like Ticketmaster.  We are skeptical about performance but would like to trust the platform and test it under a load.  If the client and server bits are running on the same machine via HTTP, should the routing be set to  localhost and not trying to go back out to https://oqtane.cmrus.com/?  We might be engaging too much of the routing and load balancing in the works.  to me it looks like the socket connection is being refused. 
![image](https://user-images.githubusercontent.com/311063/101316235-f79fa880-3810-11eb-9d90-114674edfc54.png)

**Comment by markdav-is** [comment]
@sbwalker it looks like the calls to the domain name is what's causing the issue.  it's creating a "hairpin" that is not allowed.  Is there a setting in Oqtane to make calls to the service route to localhost and not the site domain?  I think we can fix this with a local hosts entry, but it would be nice to control that thru deployment.

**Comment by sbwalker** [comment]
There is no setting which exists for this currently. ServiceBase has a method for CreateApiUrl() which all services use to make API calls. Currently this method creates relative Urls based on the Client domain. What you are asking for is a way to specify a different domain so that this method produces absolute Urls. There may also be some security implications to this change as well as the auth context needs to be passed to the domain specified and Oqtane currently relies on samesite cookies.

On a different topic, it sounds like the solution you are building is a consumer facing solution which may experience very high volume. For this scenario I would suggest that Blazor WebAssembly may be the more scalable option based on its technical characteristics.

**Comment by mixlplix** [comment]
Hi Shaun, looks like we got this figured out.  The issue was that we had an entry in the local hosts file that was directing the response back to the local machine’s IIS which does not have port 443 enabled.  When I removed the host file entry it worked!

 

Thanks for you support

 

</doug>

**Comment by sbwalker** [comment]
That is good news... I am glad you were able to resolve the configuration issue. I guess the most important conclusion from this thread is that based on the client/server architecture of Oqtane, the client always makes HTTP calls to the server API. When running on Blazor Server this essentially means you are using "hairpinning" (or NAT loopback)  where the server is accessing itself via the external IP address of the LAN/router.

**Comment by markdav-is** [comment]
One option I can think of for this that would short-circuit the http calls would be to call the controllers directly somehow as a project reference. I'll backlog that on our project and share any discoveries.

**Comment by sbwalker** [comment]
Since the goal in Oqtane is to support multiple hosting models in the same code base, Oqtane uses a standard client/server model. Theoretically it would be possible to rearchitect so that when running on Blazor Server the service classes in the Client project could interface directly with the Server project - however it could not be done via the existing Web API layer - it would need some other equivalent "manager" layer created which duplicate the API functionality. In addition it is important to note that when running on Blazor Server the DI scoping model is different which requires an alternate approach for EF Core ( https://docs.microsoft.com/en-us/aspnet/core/blazor/blazor-server-ef-core?view=aspnetcore-5.0 ) - so you could not share the same data access logic for Blazor Server and Blazor WebAssembly. Based on all of these complications, the simplest approach is to use a standard client/server model.

I am not sure if you saw my note above, but for scalability in a consumer-facing application I would seriously consider using Blazor WebAssembly over Blazor Server. Blazor WebAssembly is a true SPA architecture similar to React/Angular/Vue where the client and server only have stateless data interactions. Whereas, Blazor Server maintains a persistent stateful connection between client and server for every client and has continuous interactions for UI events in addition to data.

**Comment by markdav-is** [comment]
thanks for the details.  I think it's best for us to follow the pattern, but if we end up bottlenecking because we are too service oriented, short circuiting is a terrific way to cut down on network traffic and serialization. I agree with all your points, and this would be a hack on our part for performance.

**Comment by markdav-is** [comment]
I'm really interested in your comments around WASM and high performance.  My thought has been that WASM would be slow to load on that initial hit and thus a poor choice for a high performance site.  So we were planning to dump RAM into the webheads to account for the SignalR overhead.  I'd like to know more.  perhaps this warrants a specific discussion thread.

**Comment by jimspillane** [comment]
@markdav-is I would evaluate the [UI latency](https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-5.0#ui-latency) of the Server hosted model, especially if you have global browsers.

**Comment by sbwalker** [comment]
@markdav-is you are correct that when running on WebAssembly, the initial browser connection needs to download assemblies. However from that point on the assemblies are cached in the browser and do not impose any performance issues. A true SPA architecture should always be considerably more scalable than a traditional server architecture because a SPA is stateless and requires far less information to be transferred between the browser and the server as part of user interactions. The main performance aspects to consider for scalability is the sheer number of interactions between physically separated components in the system, the latency of those interactions, the average payload size, and whether those interactions are stateful or stateless. Both Blazor Server and Blazor WebAssembly are excellent options however you need to choose the one which is best suited to your specific business requirements.

**Comment by sbwalker** [comment]
I should also mention that the user experience in Blazor Server is not always optimal - especially for consumer facing applications. Consider the common scenario where a user has a Blazor Server application open in a browser tab and does not interact with it for a few minutes. The SignalR connection will drop and the user will see the following message:

![image](https://user-images.githubusercontent.com/4840590/101918677-a2a8bd00-3b97-11eb-887c-967ea9374691.png)

**Comment by markdav-is** [comment]
up until this point I was considering Oqtane primarily a Blazor Server solution.  Sounds like I should take another look at how to make Oqtane work in a WASM mode.  Due to the dynamic nature of the loading of modules it's going to be hard to optimize the size of the WASM payload.  There are improvements to shrinking the payload size for WASM on .net 5.0 and more coming in 6 but these are all compile-time optimizations.  Maybe Oqtane is not the best choice for the WASM piece but we could continue to build modules, use the API endpoints, etc.  I had not considered building a separate WASM client, but this might be an appropriate solution for high-volume and mobile app situations.

**Comment by sbwalker** [comment]
Oqtane already fully supports Blazor Server and Blazor WebAssembly. To run on WebAssembly you can change a single setting in Oqtane.Server\appsettings.json:

```
{
  "ConnectionStrings": {},
  "Runtime": "WebAssembly",
  ...
}
```

When running on WebAssembly, Oqtane manages the dynamic downloading assemblies to the browser. It currently does this by creating a compressed ZIP of all required assemblies (to minimize the size), sending them to the client, decompressing them, and loading them into the browser app domain.

**Comment by Rodien** [comment]
@markdav-is you guys should look at Lazy loading: https://docs.microsoft.com/en-us/aspnet/core/blazor/webassembly-lazy-load-assemblies?view=aspnetcore-5.0

@sbwalker metion lazy loading in a presentation he gave a month or so ago, so Oqtane will probably make use of lazy loading in a future upgrade.

**Comment by sbwalker** [comment]
@Rodien one thing I should mention is that the Lazy Loading enhancement that Microsoft introduced in .NET 5 has a major limitation. Real world client projects running on WebAssembly often contain service classes for interaction with back-end APIs… and developers usually want to utilize the standard .NET Core dependency injection approach. So these services need to be registered in the service container. .NET Core requires that all services must be registered during startup. However when you take advantage of Lazy Loading, it defers the assembly loading until AFTER startup which means that you are unable to register your services. The only workaround is to abandon dependency injection altogether for your services and either use HttpClient directly in your components or instantiate service classes manually ( ie. service = new Service() ) – neither of which are optimal. 

It is unfortunate that many of the new Blazor "features" in .NET 5 are not really usable in a real world application like Oqtane. The new Javascript registration feature only supports ES6 style modules - not regular JavaScript libraries ( whereas Oqtane has a solution that supports both ). CSS isolation is handled during build/deployment - which does not work in a dynamic application framework ( Oqtane has its own dynamic solution for CSS ).

**Comment by markdav-is** [comment]
We tried flipping the bits to run our project WASM style and there's lots of missing libraries.  We need to revisit how we are bundling for release and make sure we have all the dependent libraries.  When we moved our modules to .Net 5.0 we started getting release errors with missing libraries. Changes in compiling feel incompatible with Oqtane dynamic loading.  I can't see how we can get a smaller WASM size and account for anything that would be needed to load modules.   @sbwalker mentioned zipping assemblies, and I'm not sure where/how to configure that.  We have nuget packages to include.  Any links to docs on that would be much appreciated.  We are going to experiment with building a light-weight client this weekend outside of oqtane that can directly reference the modules.  We know that this experiment breaks the oqtane deployment model, but we are more comfortable playing around with the native .Net 5.0 tooling.  I'm going to try and understand the proper Oqtane WASM way as well.  We might need some guidance.  I would happily pay for a series of classes around oqtane deployments and scaling.  It's the most confusing and hard to sell part of the platform.
